### DISCUSSION TITLE
Operator Representations at the Worker / Serialization / Lowering

### DISCUSSION THREAD

**Discussion by adrianmichalke** [nebulastream/nebulastream#367] [Ideas]
## Problem
The notion of physical and logical operators is fuzzy in NebulaStream. 

Traditionally, the optimizer takes a logical plan and returns a qep (query execution plan) consisting of physical operators. The physical operators represent the executable operations to execute the query. Logical operators are only the concern of the optimizer and not of the query execution.

In our case, the optimizer is a centralized, separated component ie not part of the worker. Following this, the worker does not need to know about logical operators and just executes the qep.

Nevertheless, NebulaStream is special in that regard as we cannot assume that all optimization decisions can be done at the centralized optimizer level. Thus, we want to allow for local optimizations as well (e.g., branching v predication, use of compression, GPU usage, etc.).

In our current implementation, the NES worker receives a logical query plan from the optimizer and lowers it in various stages down to a 'query compilable' representation.

Our optimizer (as of 'old' NES) includes rewrite phases transforming logical plans to logical plans. The physical representation of an operator is determined during lowering on the worker. Current rules are quite simple.

## Alternatives
Alternatives differ in how high-level the operator representations are that are exposed to the worker. Alternatives from higher to lower:

**A1  `Centralized Logical Plan with Local Lowering` (current):** All optimizations that can happen at the logical level, e.g., operator reordering, happen at the optimizer. The optimizer returns a logical plan. The worker receives the logical plan and lowers it/optimizes it to physical operators based on local rules.

**A2  `Hint-Based Lowering at Worker`:** Logical optimizations happen at the optimizer. The worker receives logical plan and lowers it to physical plan based on local rules. Optimizer includes _lowering hints_ to logical plan to which physical operators should be lowered.  Hints have different levels of strictness in which the worker is forced to follow them.

**A3   `Shipping Physical QEP`:** Majority of optimizations happen at the optimizer. The worker receives the physical qep. The worker can apply additional local optimizations. We allow to (plugin) register optimization rules to the worker transforming physical qep to physical qep during query setup. Rules can be either internal ('pipelining', etc.) or external ('adaptive usage rule for your new custom operator') analog to notion of internal and external plugins for sources.

**A4   `Shipping Nautilus QEP`:** Worker receives the nautilus executable query plan. Optimizations happen all at the optimizer (except for small query adaptions).

Just to mention, there are more radical options for shipping even lower levels such as MLIR, and machine code.

**Pro/Con Draft:**
| Alternative | Description                              | Advantages                                                                                                                       | Disadvantages                                                                                      |
|-------------|------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------|
| **A1**      | Logical plan with local lowering         | + Low burden on central optimizer<br>                                                              | - Potential inconsistency in local optimizations                                                   |
| **A2**      | Hint-based lowering at worker            | + Balances central guidance with local flexibility<br>                                        | - Complexity in managing and interpreting hints                                                    |
| **A3**      | Shipping physical QEP                    | + High robustness<br>                                                                            | - High communication overhead<br>- Less flexibility for local optimization                       |
| **A4**      | Shipping Nautilus QEP      | + High robustness<br>+ All optimizations centralized                                                                 | - No local optimization<br>                             |

**Further readings:**
Overview of query plan representations in other systems: [https://unifiedqueryplan.github.io/](https://unifiedqueryplan.github.io/)

##  Problem
To increase robustness, and foster simplicity and extensibility, we should consider reducing abstraction layers and merging operator representations. 
This is guided by two observations: 
1) Nautilus operators and physical operators share for nearly all operators the same operator-specific attributes. Nautilus operators are already 'physical operators', as typically defined as representation of the executable operations to execute the query.
2) A protobuf-serialized operator is a operator representation itself. Also serialization and deserialization do not have to be symmetric. Instead of lowering `logical --> protobuf --> logical --> physical --> nautilus` we can simplify eg to `logical --> protobuf --> physical --> nautilus`.

## Alternatives
The following alternatives exist:
**A1  `logical --> protobuf --> logical --> physical --> nautilus` (current)**
**A2  `logical --> protobuf --> physical --> nautilus`**
**A3  `logical --> protobuf --> logical --> nautilus`**
**A4  `logical --> protobuf --> nautilus`**

## 3) Operator Extensibility

The idea is to follow the design of DD [20240701_static_plugin_registry.md](https://github.com/nebulastream/nebulastream-public/blob/main/docs/design/20240701_static_plugin_registry.md) by @alepping. We only outline the specific considerations for operators here.

Similar to source plugins, there are attributes shared by all operations and operator-specific attributes:

- Shared attributes: OperatorId

- Operator-specific attributes: arbitrary 'basic types'. Implemented as unordered_map<string, variant> with variant of int, uint, float, double, char, and string.

In addition to basic data types, operators may require expressions as part of their configuration eg, predicates in filter operators, or expressions in map operators. As expression representation, we extend the variant type in the attribute map to include an `expression` type containing a string representation of the expression (for expression plugin lookup) and a configuration (unordered_map<string, variant>) for expression-specific attributes.

**Comment by keyseven123** [comment]
# Logical and Physical Operators at the Worker
- I like A2 or A3 as it gives the worker some flexibility while also given the user the option to force a specific optimization/lowering 
	- For A2: Can we also here allow plugins to register specific optimization rules both on the coordinator and on the worker?
	- The main difference for A2 vs. A3 is what we ship/send over the network from the coordinator to the worker. 
		- Is shipping physical easier than logical? 
		- What about, if we have a distributed coordinator? In that sense shipping logical operators would be easier, as we only have to implement one serialization of protobuf <---> logical.
		
- I am against A4 
	- as this will lead to no opportunity to do anything on the worker, in terms of optimizations.
	- additionally, we are limiting ourselves to nautilus as our backend, which is an unnecessary limitation

**Comment by keyseven123** [comment]
# Simplify Operator Representations

I would advocate for doing no optimizations on the nautilus operators. I think that we can merge the current logical and physical operators into a logical operators. What I do not want is to have a query plan of nautilus operators that we then do some optimizations. The nautilus operators should focus on the actual execution of the operator over some tuple buffer and not being used to do some optimizations. 

For me there are also certain other constraints. We should have clear interfaces and one component should do on job well, e.g., `logical encoded operators in  protobuf --> logical` is one job. `logical encoded operators in protobuf ---> physical operators` are two IMHO. As the lowering can be more complex than simply going from one logical operator to one physical operator. For example, a join operator consists of at least three executable operators. Furthermore, we might want to introduce sub-operators into our execution to more efficiently use accelerators. Such a change should not touch the deserialization IMO.

**Comment by alepping** [comment]
# Logical -> Protobuf -> Nautilus (LPN)
## Assumptions
A1: Nautilus is a good physical representation
- in other systems, such as DuckDB, the physical representation is the actual operator implementation, that is, e.g., the concrete hash join, which can be executed. Thus Nautilus is not an unordinary physical representation

A2: We don't need to overload the Nautilus Operators
- we can write a class that performs transformations on nautilus query graphs, e.g., pipelining, or splitting a Join into the three different concrete operators

A3: We develop a prototype that shows that this approach is technically feasible
- without this prototype, we won't start implementing this approach. If we discover that it is not feasible, I would suggest going back to Logical -> Protobuf -> Logical -> Nautilus

A4: We describe how we can extend the LPN approach
- the prototype shows how pipelining and handling the join works and demonstrates how this design can be extended, potentially to make decisions concerning memory layout, or whether to compress data
- with the right design adding a separate representation between protobuf and nautilus should not be feasible, even with limited resources

## Goals
G1: reduce complexity (limit implementation to actual current requirements)
G2: design with (potential) future requirements in mind (extensibility must be feasible)

## Solution (very rough sketches)
S1 (G1): By using the Logical -> Protobuf -> Nautilus (LPN) approach, we reduce lowering to only essential pieces. Additionally, we create a clear separation between the Coordinator/Optimizer -Logical Query Plan- and the Worker -Nautilus Query Plan-. Thus, logical operators clearly belong to the optimizer and nautilus operators clearly belong to the worker. Protobuf sits in between (and can be swapped out). If we discover that performing pipelining and splitting up join operators, etc. is to complex on a nautilus query plan, we abort the prototype and fallback to another approach, e.g. Logical -> Protobuf -> Logical -> Nautilus.
S2 (G2): If we can perform pipelining and operator splitting, e.g., Join, on the nautilus query plan (hard requirement of prototype) we should be able to perform other modifications too, without requiring changes in the nautilus operator implementations. If we discover that we want to perform modifications that are very difficult to implement with the LPN approach, we can fallback, by adding back a logical layer on top of Nautilus.\
Plugins for operators depend on both nautilus operators (worker) and logical operators (coordinator/optimizer) and need to implement (at least) two different registry interfaces. As long as the both types of operators are available in one repository, it should not be a problem to develop (external) plugins.\
The optimizer can still enrich the logical query plan with hints. Either during serialization, or an a dummy query plan, the worker can utilize these hints.

## Problems with LPN
P1: if nautilus is our physical representation, an optimizer that takes a logical query plan and produces a physical query plan (most optimizers do that) would need to produce a nautilus query plan and would therefore need to depend on nautilus. Otherwise the optimizer needs to produce a logical query plan.
P2: we would need to create a nautilus operator registry and each nautilus operator would need to register in that registry. If we implement validation functions and potentially other functions for the registry. This potentially is a significant investment.
P3: we either push complexity into the deserialization, e.g., splitting up a logical join in the three nautilus operators and realizing optimizer hints, or we create dummy operators, e.g., a dummy nautilus join, which we later split. Both solutions are not ideal.

**Reply by alepping** [reply]
P4: we are more tightly coupled to GRPC/Protobuf

**Comment by zeuchste** [comment]
A2: The question is of course why and how can the optimizer create hints? On what did he bases the hints, I guess part of the answer is that the optimizer sees the entire plan (also the upstream and downstream part of the QEP) and in addition maybe also the other plans that are running and in addition also the overall topology and then the question is if he also sees statistics over the toplogy like path occupation. In general my point would be, on what do we actually base which decision. There are several categories:
	- general rewrite rules on the logical plan that are always benefical like selection push down
	- rewrite rules on the logical plan that require statistics like reordering
	- local rewrite rules that take the current load of the system into accout e.g. use NL join because no memory is left
	- local rewrite rules that use specific properties of the worker like there is a GPU and AVX512

In general the "normal" definition of a physical QEP does not hold in our system anymore and I think we need a new one because we have more levels:
	- Logical level: join, seleciton, map
	- algorithm (formally kown as physical): here algorithm selection takes plase so NL join or Hash join
	- hardware: here specical hardware opportunites have to be taken into account
	- load: here the current load has to be taken into account
I would argue logical is defnitely on the optimizer and hardware and load defnitely on the worker the question is now what is with algorithm, there we could make a both

In addition I would add Dwi here to the discussion, as the new "common" way is to lower even further to sub-operators like the work of TU Munich. In addition you mention nautilius as operators how do they group in here?

I like the idea of a optimization registry but again we need to find out what do we need for every optimization and where it should be applied.

**Reply by keyseven123** [reply]
So for me, Nautilusoperators are sub-operators. I did not want to open-up this bag of worms of sub-operators. 
Therefore, I do not think that we should go directly to nautilus operators, as we are creating a lot of hassle if we do any optimizations in terms of splitting this physical join operator into three or more sub-operators. For example, we might want to change the order of different steps so that we can move steps onto accelerators that benefit.
